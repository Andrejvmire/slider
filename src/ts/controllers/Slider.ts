import SliderModel from "../models/SliderModel";
import SliderView from "../views/SliderView";

export default class Slider implements ISubscriber {
    private _model: ISlider;
    private _view: IViewPublisher;

    constructor(options: MainOptionsType, parent: JQuery) {
        this._model = new SliderModel({
            ...options
        });
        this._view = new SliderView({points: [60,257], ruler: [20, 400], tooltip: true}, parent);
        this._model.attach(this);
    }

    update(data?: any): void {
    }
}