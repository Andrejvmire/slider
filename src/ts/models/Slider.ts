import Point from "./Point";

type returnEmptyObject = {};
type returnValue = {
    from: number,
    to: number,
    points: number[]
}

export class Slider {
    private _points: Point[] | undefined;

    private pointInRange(value: number): boolean {
        if (value >= this._from && value <= this._to) return true;
        throw new Error(`Point value ${value} out of range: less ${this._from} or more ${this._to}`);
    }

    constructor(private _from: number, private _to: number, private _step: number = 1) {
    }

    points(points: [number]): Slider;
    points(points: [number, number]): Slider;
    points(points: any[]): Slider {
        points = (points[0] === points[1]) ? [points[0]] : points;
        try {
            this._points = points
                .filter(
                    item => this.pointInRange(item)
                )
                .map(
                    item => new Point(item)
                )
        } catch (e) {
            throw new Error(e.message);
        }
        return this;
    }

    values(): returnValue | returnEmptyObject {
        if (typeof this._points === 'undefined') return {};
        return {
            from: this._from,
            to: this._to,
            points: this._points
                .sort((a, b) => a.value - b.value)
                .map(
                    item => item.value
                )
        }
    }

    move(to: number): Slider {
        try {
            this.pointInRange(to);
        } catch (e) {
            return this;
        }
        if (typeof this._points === "undefined") this._points = [new Point(to)];
        this._points
            .reduce(
                (prevValue, curValue) => {
                    return (Math.abs(prevValue.value - to) < Math.abs(curValue.value - to)) ? prevValue : curValue;
                }
            )
            .value = to;
        return this;
    }

    private from(currentValue: number, to: number): Slider {
        if (typeof this._points === 'undefined') return this;
        try {
            this.pointInRange(currentValue + to)
        } catch (e) {
            return this;
        }
        this._points
            .reverse()
            .map(item => {
                if (item.value === currentValue) {
                    item.value += to;
                }
            })
        return this;
    }

    nextPoint(currentValue: number): Slider {
        return this.from(currentValue, this._step);
    }

    prevPoint(currentValue: number): Slider {
        return this.from(currentValue, -this._step);
    }
}