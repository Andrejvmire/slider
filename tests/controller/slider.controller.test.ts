import Slider from "../../src/ts/controllers/Slider";
import $ from 'jquery';

describe("Тестируем контроллер", function () {
    let slider: ISubscriber & IPublisher;
    const options: MainOptionsType = {
        ruler: [-2399, 5484],
        points: [3, 1688]
    };
    const parent = $(document.createElement('div'));
    const subscriber = {
        update: jest.fn(x => x)
    }
    beforeEach(() => {
        slider = new Slider(options, parent);
        slider.attach(subscriber);
    });
    it('должен соответствовать снимку', function () {
        expect(parent)
            .toMatchSnapshot()
    });
    it('должен переместить точку на новое место, оповестить подписчика и соответствовать снимку', function () {
        slider.update([-400]);
        expect(slider.state)
            .toStrictEqual([-400, 1688]);
        expect(subscriber.update.mock.results[0].value)
            .toStrictEqual([-400, 1688]);
        expect(parent)
            .toMatchSnapshot();
    });
})