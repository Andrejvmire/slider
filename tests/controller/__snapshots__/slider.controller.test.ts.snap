// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Тестируем контроллер должен переместить точку на новое место, оповестить подписчика и соответствовать снимку 1`] = `
jQuery {
  "0": <div
    class="slider slider__container slider__container_horizontal"
  >
    <div
      class="slider slider__ruler slider__ruler_horizontal"
    >
      <div
        class="slider slider__ranger"
        style="left: 30.470633007738172%; width: 21.375110998350877%;"
      />
      <span
        class="slider slider__point slider__point_horizontal"
        style="left: 30.470633007738172%;"
      >
        <span
          class="slider slider__tooltip slider__tooltip_horizontal"
        >
          3
        </span>
      </span>
      <span
        class="slider slider__point slider__point_horizontal"
        style="left: 51.84574400608905%;"
      >
        <span
          class="slider slider__tooltip slider__tooltip_horizontal"
        >
          1688
        </span>
      </span>
    </div>
    <div
      class="slider slider__ruler slider__ruler_horizontal"
    >
      <div
        class="slider slider__ranger"
        style="left: 25.35836610427502%; width: 26.48737790181403%;"
      />
      <span
        class="slider slider__point slider__point_horizontal"
        style="left: 25.35836610427502%;"
      >
        <span
          class="slider slider__tooltip slider__tooltip_horizontal"
        >
          -400
        </span>
      </span>
      <span
        class="slider slider__point slider__point_horizontal"
        style="left: 51.84574400608905%;"
      >
        <span
          class="slider slider__tooltip slider__tooltip_horizontal"
        >
          1688
        </span>
      </span>
    </div>
  </div>,
  "length": 1,
}
`;

exports[`Тестируем контроллер должен соответствовать снимку 1`] = `
jQuery {
  "0": <div
    class="slider slider__container slider__container_horizontal"
  >
    <div
      class="slider slider__ruler slider__ruler_horizontal"
    >
      <div
        class="slider slider__ranger"
        style="left: 30.470633007738172%; width: 21.375110998350877%;"
      />
      <span
        class="slider slider__point slider__point_horizontal"
        style="left: 30.470633007738172%;"
      >
        <span
          class="slider slider__tooltip slider__tooltip_horizontal"
        >
          3
        </span>
      </span>
      <span
        class="slider slider__point slider__point_horizontal"
        style="left: 51.84574400608905%;"
      >
        <span
          class="slider slider__tooltip slider__tooltip_horizontal"
        >
          1688
        </span>
      </span>
    </div>
  </div>,
  "length": 1,
}
`;
